apollo-runtime/src/main/java/com/apollographql/apollo/internal/batch/BatchPoller.kt

```diff
@@ -0,0 +1,76 @@
+package com.apollographql.apollo.internal.batch
+
+import com.apollographql.apollo.api.internal.ApolloLogger
+import com.apollographql.apollo.exception.ApolloException
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.Executor
+import java.util.concurrent.Executors
+import java.util.concurrent.ScheduledExecutorService
+import java.util.concurrent.ScheduledFuture
+import java.util.concurrent.TimeUnit
+
+/**
+ * Class responsible for holding the queue of requests, and execute them as a batch every X ms,
+ * as defined by the BatchConfig.
+ */
+class BatchPoller(
+    private val batchConfig: BatchConfig,
+    private val dispatcher: Executor,
+    private val batchHttpCallFactory: BatchHttpCallFactory,
+    private val logger: ApolloLogger,
+    private val periodicJobScheduler: PeriodicJobScheduler = PeriodicJobSchedulerImpl()
+) {
+
+  private val queryQueue: ConcurrentLinkedQueue<QueryToBatch> = ConcurrentLinkedQueue()
+
+  fun start() {
+    stop()
+    periodicJobScheduler.schedulePeriodicJob(
+        initialDelay = 0,
+        interval = batchConfig.batchIntervalMs,
+        unit = TimeUnit.MILLISECONDS
+    ) {
+      maybeExecuteBatchQuery()
+    }
+  }
+
+  fun stop() {
+    periodicJobScheduler.cancel()
+  }
+
+  fun enqueue(query: QueryToBatch) {
+    if (!periodicJobScheduler.isRunning()) {
+      throw ApolloException("Trying to batch queries without calling ApolloClient.startBatchPoller() first")
+    }
+    val newQueueSize = synchronized(this) {
+      queryQueue.add(query)
+      logger.d("Enqueued Query: ${query.request.operation.name().name()} for batching")
+      queryQueue.size
+    }
+    // When the amount of queries in the queue reaches the max batch size, trigger the HTTP request without waiting for the batch interval
+    if (newQueueSize >= batchConfig.maxBatchSize) {
```

Comment:
Nit: there's a small chance that a new query is enqueued from another thread just before or after this check happens so `maybeExecuteBatchQuery()` will be called twice and the first time, might send `maxBatchSize + 1` instead of just `maxBatchSize`. I don't think this should be an issue? If it is, I guess there's no way around holding the lock around the enqueue + send with something like `enqueueAndMaybeExecuteBatchQuery`

Reply:
yeah I see what you mean. I could make `maybeExecuteBatchQuery()` not synchronized, so I can use it inside the synchronize block above, and synchronize at the call site in the periodic callback as well. What do you think?

Reply:
Sounds good to me 👍